### PR TITLE
Add IsRef property to IConditionalOperation and ISimpleAssignmentOperation, add RefKind property to ILocalSymbol.

### LINKED ISSUE #21311: Design Question : Should we have a special node for ref (C# 7.0 feature)
Out Var https://github.com/dotnet/roslyn/issues/18303 
Patterns  https://github.com/dotnet/roslyn/issues/19927 
Tuple https://github.com/dotnet/roslyn/issues/10856

### PR DESCRIPTION
**Pull request by AlekseyTs**
Closes #21311.

vso : https://devdiv.visualstudio.com/DevDiv/NET%20Developer%20Experience%20Productivity/_workitems/edit/517771

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler, @dotnet/analyzer-ioperation, @mavasani, @cston Please review 15.5 IOperation fix.

**Comment by AlekseyTs** [comment]
CC @jaredpar, @gafter, @AnthonyDGreen For public API change on ILocalSymbol.

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs [hunk 1]
```diff
@@ -183,6 +183,12 @@ public override void VisitLocal(ILocalSymbol symbol)
             {
                 AddKeyword(SyntaxKind.RefKeyword);
                 AddSpace();
+
+                if (symbol.RefKind == RefKind.RefReadOnly)
+                {
+                    AddKeyword(SyntaxKind.ReadOnlyKeyword);
+                    AddSpace();
+                }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs]: Should the `SymbolDisplay` change be a separate PR, perhaps post 15.5?

File: src/Compilers/Core/Portable/Operations/IConditionalOperation.cs [hunk 2]
```diff
@@ -31,6 +31,11 @@ public interface IConditionalOperation : IOperation
         /// Operation to be executed if the <see cref="Condition"/> is false.
         /// </summary>
         IOperation WhenFalse { get; }
+
+        /// <summary>
+        /// Is result a managed reference
```
Comment (tmat) [inline src/Compilers/Core/Portable/Operations/IConditionalOperation.cs]: >Is result a managed reference [](start = 12, length = 29)

Is this not encoded in the type of the operation?

File: src/Compilers/Core/Portable/Symbols/ILocalSymbol.cs [hunk 3]
```diff
@@ -23,11 +23,17 @@ public interface ILocalSymbol : ISymbol
         bool IsConst { get; }
 
         /// <summary>
-        /// Returns true if this local is a ref-local.
+        /// Returns true if this local is a ref local or a ref readonly local.
+        /// Use <see cref="RefKind"/> to get more detailed information.
         /// </summary>
         bool IsRef { get; }
```
Comment (tmat) [inline src/Compilers/Core/Portable/Symbols/ILocalSymbol.cs]: Why keep this property if the same information is in RefKind?

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.cs]: > Should the SymbolDisplay change be a separate PR, perhaps post 15.5?

Do you have concerns about this change? It feels low risk to me.

File: src/Compilers/Core/Portable/Operations/ISimpleAssignmentOperation.cs [hunk 4]
```diff
@@ -16,6 +16,10 @@ namespace Microsoft.CodeAnalysis.Operations
     /// </remarks>
     public interface ISimpleAssignmentOperation : IAssignmentOperation
     {
+        /// <summary>
+        /// Is this a ref assignment
+        /// </summary>
+        bool IsRef { get; }
```
Comment (tmat) [inline src/Compilers/Core/Portable/Operations/ISimpleAssignmentOperation.cs]: >bool IsRef { get; } [](start = 8, length = 19)

Why not expose RefKind property instead?

File: src/Compilers/Core/Portable/Operations/IConditionalOperation.cs [hunk 5]
```diff
@@ -31,6 +31,11 @@ public interface IConditionalOperation : IOperation
...
+        /// </summary>
+        bool IsRef { get; }
```
Comment (tmat) [inline src/Compilers/Core/Portable/Operations/IConditionalOperation.cs]: >bool IsRef { get; } [](start = 8, length = 19)

Expose RefKind property instead?

File: src/Compilers/Core/Portable/Symbols/ILocalSymbol.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Symbols/ILocalSymbol.cs]: > Why keep this property if the same information is in RefKind?

We already shipped it, cannot take it back.

Comment (tmat) [reply src/Compilers/Core/Portable/Symbols/ILocalSymbol.cs]: I see. Ok.

---
In reply to: [148093610](https://github.com/dotnet/roslyn/pull/22933#discussion_r148093610) [](ancestors = 148093610)

File: src/Compilers/Core/Portable/Operations/ISimpleAssignmentOperation.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ISimpleAssignmentOperation.cs]: > Why not expose RefKind property instead?

Because the kind of the reference is irrelevant here. Even compiler doesn't capture the kind.

File: src/Compilers/Core/Portable/Operations/IConditionalOperation.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/IConditionalOperation.cs]: > Expose RefKind property instead?

Because the kind of the reference is irrelevant here.

File: src/Compilers/Core/Portable/Operations/ISimpleAssignmentOperation.cs (hunk 4 above)
Comment (tmat) [reply src/Compilers/Core/Portable/Operations/ISimpleAssignmentOperation.cs]: OK. Thanks

---
In reply to: [148094214](https://github.com/dotnet/roslyn/pull/22933#discussion_r148094214) [](ancestors = 148094214)

**Review by cston — APPROVED** [review]

**Review by mavasani — APPROVED** [review]

**Review by agocke — APPROVED** [review]
LGTM

File: src/Compilers/CSharp/Portable/Symbols/LocalSymbol.cs [hunk 6]
```diff
@@ -321,9 +321,9 @@ public object ConstantValue
 
         internal abstract ImmutableArray<Diagnostic> GetConstantValueDiagnostics(BoundExpression boundInitValue);
 
-        public bool IsRef => RefKind == RefKind.Ref;
+        public bool IsRef => RefKind != RefKind.None;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/LocalSymbol.cs]: Consider updating some compiler tests for this.

**Comment by jcouv** [comment]
>                 Return False

Should this be proactively updated?

---
Refers to: src/Compilers/VisualBasic/Portable/Symbols/Source/LocalSymbol.vb:285 in b70c08e. [](commit_id = b70c08e27aa90fe86e21de08bd7b31581d689ad0, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/LocalSymbol.cs (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/LocalSymbol.cs]: I added new tests along with symbol display.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 7]
```diff
@@ -1232,6 +1238,12 @@ public override void VisitArrayInitializer(IArrayInitializerOperation operation)
         public override void VisitSimpleAssignment(ISimpleAssignmentOperation operation)
         {
             LogString(nameof(ISimpleAssignmentOperation));
+
+            if (operation.IsRef)
+            {
+                LogString(" (IsRef)");
```
Comment (jcouv) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: I didn't see which test exercises this.

**Review by jcouv — APPROVED** [review]
LGTM with one test suggestion.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs (hunk 7 above)
Comment (AlekseyTs) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: > I didn't see which test exercises this.

None at the moment. Only lowered tree can have nodes like this. Once ref reassignment feature is supported, then we can test this API. 

---
In reply to: [148105993](https://github.com/dotnet/roslyn/pull/22933#discussion_r148105993) [](ancestors = 148105993)

**Comment by AlekseyTs** [comment]
>                 Return False

VB doesn't support ref locals.

---
In reply to: [340882949](https://github.com/dotnet/roslyn/pull/22933#issuecomment-340882949) [](ancestors = 340882949)

---
Refers to: src/Compilers/VisualBasic/Portable/Symbols/Source/LocalSymbol.vb:285 in b70c08e. [](commit_id = b70c08e27aa90fe86e21de08bd7b31581d689ad0, deletion_comment = False)

File: src/Compilers/Core/Portable/Symbols/ILocalSymbol.cs [hunk 8]
```diff
@@ -23,11 +23,17 @@ public interface ILocalSymbol : ISymbol
         bool IsConst { get; }
 
         /// <summary>
-        /// Returns true if this local is a ref-local.
+        /// Returns true if this local is a ref local or a ref readonly local.
+        /// Use <see cref="RefKind"/> to get more detailed information.
         /// </summary>
```
Comment (VSadov) [inline src/Compilers/Core/Portable/Symbols/ILocalSymbol.cs]: So, "IsRef" is true for readonly references as well. 
I do not have a strong opinion on this, but we have similar API for methods/properties, should we make sure it is also "inclusive"?

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Symbols/ILocalSymbol.cs]: > So, "IsRef" is true for readonly references as well.

Yes.

> I do not have a strong opinion on this, but we have similar API for methods/properties, should we make sure it is also "inclusive"?

IMethodSymbol and IPropertySymbol also have ReturnsByRefReadonly property, I'll let you decide what correlation should be between it and ReturnsByRef property. If any adjustments are needed, I think they are out of scope of this PR.

**Comment by jinujoseph** [comment]
pre-approved to merge via [link](https://devdiv.visualstudio.com/DevDiv/NET%20Developer%20Experience%20Productivity/_workitems/edit/517771)
